FuelLabs/fuel-indexer
# enhancement: Update --replace-indexer and add --remove-data V2

### Description

A redo of https://github.com/FuelLabs/fuel-indexer/pull/1146 but with the kill notification mechanism removed.

### Testing steps

#### This PR adds an automatic CI test:

https://github.com/FuelLabs/fuel-indexer/pull/1225/files#diff-61710fa718a25e0e49156f4ac2682a97e9eb0d9fb5cb5316f09f924dde896982R307

#### Manual testing:

1. Start the `fuel-indexer`:

```
cargo run -p fuel-indexer -- run --run-migrations --fuel-node-host beta-3.fuel.network --fuel-node-port 80
```

2. Deploy an indexer:

```
cargo run --release -p forc-index -- deploy --path examples/fuel-explorer/fuel-explorer
```

Expected result: indexer processing blocks.

3. Redeploy an indexer, and keep the data:

```
cargo run --release -p forc-index -- deploy --path examples/fuel-explorer/fuel-explorer --replace-indexer
```

Expected result: indexer is replaced, and it resumes from where `fuel-explorer` left off:

```
INFO fuel_indexer::service: 397: Resuming Indexer(fuellabs.explorer) from block 1760
```

4. Redeploy an indexer, and remove the data:

```
cargo run --release -p forc-index -- deploy --path examples/fuel-explorer/fuel-explorer --replace-indexer --remove-data
```

Expected result: the indexer restarts from block 1:

```
INFO fuel_indexer::service: 397: Resuming Indexer(fuellabs.explorer) from block 1
```

5. Modify the schema, and attempt to redeploy the indexer without removing the data:

```
echo "type Ty { id: ID! }" >> examples/fuel-explorer/fuel-explorer/schema/fuel_explorer.schema.graphql 
```

```
chmod +x ./scripts/utils/build_modules.bash && ./scripts/utils/build_modules.bash
```

```
cargo run --release -p forc-index -- deploy --path examples/fuel-explorer/fuel-explorer --replace-indexer
```

Expected result: an error:

```
▹▹▹▹ 🚀 Deploying...
{
  "details": "Conflict: Indexer(fuel.explorer)'s schema has changed. Use --remove-data to replace the indexer.",
  "success": "false"
}
```

6. Redeploy the modified indexer and remove old data:

```
cargo run --release -p forc-index -- deploy --path examples/fuel-explorer/fuel-explorer --replace-indexer
```

Expected result: indexer restarts from block 1.

```
INFO fuel_indexer::service: 397: Resuming Indexer(fuellabs.explorer) from block 1
```

### Changelog
- Instead of the kill notification mechanism and removing/restarting the indexer in a single request, in this PR, `forc index deploy` first issues the `remove` command and then deploys the indexer as a separate command.
- The indexer commits no data when the kill switch is flipped on—it reverts the transaction. This is done in `handle_events`, which now takes `Arc<AtomicBool>`. This allows the indexer to shutdown gracefully and prevents any data from entering the DB after the tables have been removed and recreated for the replacement indexer.

Thoughts?
